### DISCUSSION TITLE
How to typehint a generic decorator that can be applied multiple times and checks argument? Possible `@overload` bug.

### DISCUSSION THREAD

**Discussion by MaPePeR** [microsoft/pyright#7404] [Q&A] [answered]
I'm trying to add typehints to a decorator function that is used like this:

```python
T = TypeVar('T')

def register(evtype: Type[T]): #What is the proper return type?
    def decorator(handler):
        # register handler for event type
        return handler
    return decorator

class A: ...
class B: ...
class C: ...

@register(A)
def handle_a(ev: A): ...

handle_a(A())

@register(A)
@register(B)
# ... Should support infinite amount of @register calls
def handle_ab(ev: A|B): ... 

handle_ab(A())
handle_ab(B())

#Expected error cases because of wrong types:
@register(A)
def handle_b(ev: B): ... 

handle_a(B())
handle_ab(C())
```

In [this StackOverflow Question](https://stackoverflow.com/q/78101969/2256700) and [this answer](https://stackoverflow.com/a/78101970/2256700) I go through multiple iterations.

The closest solution I could find was:
```python
Callable[[Union[Callable[[T|T2], None],Callable[[T], None]]], Callable[[T|T2], None]]
```
This seems to handle all the decorator cases but does not work for the calling `handle_a(B())` case.

I tried to solve this by creating `@overload`s instead of a single Union, because I thought the "unbound" `T2` in the return type might cause trouble:
```python
@overload
def register4(evtype: Type[T]) -> Callable[[Callable[[T | T2], None]],Callable[[T | T2], None]]: ...

@overload
# Overload 2 for "register2" will never be used because its parameters overlap overload 1
def register4(evtype: Type[T]) -> Callable[[Callable[[T], None]], Callable[[T], None]]: ...
```
But that results in a warning, that the overloads overlap ... I see that the function parameters overlap, but I get the same error if I apply the `@overload` to the inner `decorator` function, so the parameters are different - even if it is a toplevel function:
```python
@overload
def decorator(handler: Callable[[T|T2], None]) -> Callable[[T|T2], None]: ...
@overload
#error: Overload 2 for "decorator" will never be used because its parameters overlap overload 1
def decorator(handler: Callable[[T], None]) -> Callable[[T], None]: ...
def decorator(handler): ...
```

And i think I can kind of show, that these are not overlapping, at least not completely, because the first `@overlap` does not accept functions as argument which argument type is not a Union:
```python
def decoratorA(handler: Callable[[T|T2], None]) -> Callable[[T|T2], None]: ...
def decoratorB(handler: Callable[[T], None]) -> Callable[[T], None]: ...

def handle_a(ev: A): ...
def handle_ab(ev: A|B): ...
decoratorA(handle_a) # error: Argument of type "(ev: A) -> None" cannot be assigned to parameter "handler" of type "(T@decoratorA | T2@decoratorA) -> None" in function "decoratorA"
decoratorA(handle_ab)

decoratorB(handle_a)
decoratorB(handle_ab)
```

I'd expect, that if the first `@overlap` would cause a "cannot be assigned" if it was a normal declaration, then it should check the next `@overlap`  declaration. So is this a bug or am I doing/expecting something wrong? Or is there a better way to add types to that `register` decorator?

This is with pyright `1.1.352`

This question is kind of similar to https://github.com/microsoft/pyright/discussions/5911, but I don't really want to alter the decorator argument, but only require, that it can handle the type specified as a decorator argument.

**Comment by Azureblade3808** [comment, marked as answer]
```python
from typing import Callable, Protocol, TypeVar, overload

T_co = TypeVar("T_co", covariant=True)

T0 = TypeVar("T0")
T1 = TypeVar('T1')

class RegisterResult(Protocol[T_co]):
    @overload
    def __call__(self, handler: Callable[[T_co | T1], None]) -> Callable[[T_co | T1], None]: ...

    @overload
    def __call__(self, handler: Callable[[T_co], None]) -> Callable[[T_co], None]: ...

def register(evtype: type[T0]) -> RegisterResult[T0]: ...

class A: ...
class B: ...
class C: ...

@register(A)
def handle_a(ev: A): ...

handle_a(A())

@register(A)
@register(B)
# ... Should support infinite amount of @register calls
def handle_ab(ev: A|B): ... 

handle_ab(A())
handle_ab(B())

#Expected error cases because of wrong types:
@register(A)
def handle_b(ev: B): ... 

handle_a(B())
handle_ab(C())
```

Edit
---
Made more similar to original code.

**Reply by MaPePeR** [reply]
Amazing. Thank you very much. This seems to pass all my tests.

If want and you have a StackOverflow you could also post your answer [here](https://stackoverflow.com/q/78101969/2256700), so I can mark it as accepted there, too.
